src/main/scala/is/hail/HailContext.scala

```diff
@@ -422,67 +423,64 @@ class HailContext private(val sc: SparkContext,
       info(s"Using sample and global annotations from ${ inputs(0) }")
   }
 
-  def readMetadata(file: String): (VariantMetadata, Boolean) = VariantDataset.readMetadata(hadoopConf, file)
+  def read(file: String, dropSamples: Boolean = false, dropVariants: Boolean = false): VariantSampleMatrix[_] =
+    readAll(List(file), dropSamples, dropVariants)
 
-  def readAllMetadata(files: Seq[String]): Array[(VariantMetadata, Boolean)] = files.map(readMetadata).toArray
-
-  def read(file: String, dropSamples: Boolean = false, dropVariants: Boolean = false,
-    metadata: Option[Array[(VariantMetadata, Boolean)]] = None): VariantDataset =
-    readAll(List(file), dropSamples, dropVariants, metadata)
-
-  def readAll(files: Seq[String], dropSamples: Boolean = false, dropVariants: Boolean = false,
-    metadata: Option[Array[(VariantMetadata, Boolean)]] = None): VariantDataset = {
+  def readAll(files: Seq[String], dropSamples: Boolean = false, dropVariants: Boolean = false): VariantSampleMatrix[_] = {
     val inputs = hadoopConf.globAll(files)
     if (inputs.isEmpty)
       fatal("arguments refer to no files")
 
-    val (mdArray, pqgtArray) = metadata match {
-      case Some(data) => data.unzip
-      case _ => inputs.map(VariantDataset.readMetadata(sc.hadoopConfiguration, _)).unzip
+    val vsms = inputs.map { input =>
+      VariantSampleMatrix.read(this, input, dropSamples = dropSamples, dropVariants = dropVariants)
     }
 
-    val vdses = inputs.zipWithIndex.map { case (input, i) =>
-      VariantDataset.read(this, input, mdArray(i), pqgtArray(i),
-        dropSamples = dropSamples, dropVariants = dropVariants)
-    }
+    checkDatasetSchemasCompatible(vsms, inputs)
 
-    checkDatasetSchemasCompatible(vdses, inputs)
+    if (vsms.length == 1)
+      return vsms(0)
 
-    vdses(0).copy(rdd = sc.union(vdses.map(_.rdd)).toOrderedRDD)
+    // I can't figure out how to write this with existentials -cs
```

Comment:
Ah, so I guess the way to do it would be for that method to return `(Array[VariantSampleMatrix[T]], ClassTag[T]) forSome {type T}`, but calling `sc.union` is still a mess because existential types are a pain to work with (I think we'd need another universally quantified function). I'm not inclined to change anything here.